### DISCUSSION TITLE
Why mesh_rag embeds two on-device models — the cactus_embed -2 forcing function

### DISCUSSION THREAD

**Discussion by alycda** [alycda/DittoXCactus#7] [Show and tell]
The app runs **two Qwen models on every phone**: `qwen3-0.6` for embeddings and `qwen3-1.7` for completion. You may be asking why we don't run one model for both — it would halve the cold-load weight, simplify boot, and free up RAM. The answer is interesting because **we didn't choose two models. The Cactus runtime forced us into it.**

## What we tried first

Initial plan: one model, both roles. The thinking was straightforward — Qwen 2.5's chat-tuned models have a transformer trunk, you can feed the same trunk an embedding query and a generation prompt, the output of one is the input of the other. Conceptually clean.

So the first attempt swapped both roles onto the bigger model:

> `ko` — "swap completion+embedding model: qwen3-0.6 → qwen3-1.7" (2026-05-23)

Boot succeeded, model downloaded, and then on-device validation surfaced this in logcat:

```
Generating embedding for text: the solar system. The sun is a yellow dwarf star, ...
Received embedding result code: -2
boot error: Bad state: cactus_embed failed for text len=218
```

Code `-2` from `cactus_embed` with no further context. The same model that completed text just fine refused to embed.

## What was actually happening

Cactus's **chat-tuned LM slugs ship without an embedding head**. The `qwen3-1.7` slug is packaged for completion: the transformer trunk plus a language-modeling head that projects hidden states to vocab logits. There's no pooling layer wired up, no similarity-tuned projection, no embed entry point. The Flutter SDK exposes both `complete()` and `embed()` against the same `CactusLM` object, but `embed()` returns `-2` when the slug doesn't carry the pieces it needs.

Cactus does ship a purpose-built embedding slug — `qwen3-embedding-0.6`, similarity-tuned, with the right pooling — and the obvious fix was to swap the embedding model to that:

> `vwn` — "fix(cactus): split embedding to qwen3-embedding-0.6 — qwen3-1.7 has no embedding head" (2026-05-23)

That commit refactored `CactusService` from one shared `_lm` to two `CactusLM` instances (`_completionLm`, `_embeddingLm`), each with its own download progress and init phase. Public API picked up `completionSlug` + `embeddingSlug` getters in place of the singular `modelSlug`. The cost was a second model download on first launch (~0.6B params on top of the 1.7B completion model), which the U14 cold-load lever later mitigated via pre-baked embeddings shipped in the seed JSON.

## Then a second forcing function

The Flutter SDK couldn't resolve `qwen3-embedding-0.6`:

> `mqu` — "fix(cactus): fall back embedding slug to qwen3-0.6 — Flutter SDK can't resolve qwen3-embedding-0.6" (2026-05-23)

So the embedder we ended up shipping is a **chat-tuned model used as an embedder anyway**. It works — the same trunk produces usable embeddings via the same internal path the SDK exposes — but it's not similarity-tuned the way the purpose-built slug would have been. This is plausibly part of why we see the cosine score distributions we do: the 0.3 threshold + entity-overlap gate are tuned for *this specific* embedder's distribution, not some Platonic ideal.

## What this means for the writeup

The two-model architecture is **a small SDK story, not a thesis story**. The mesh-RAG thesis ("vector indexes want to be CRDTs") doesn't care how many models are on each phone — what's load-bearing is that retrieval is a pure function over the merged set. The two-model split is what the runtime makes you do this week, not what the thesis demands.

But it has a real consequence for the U16 alternate-model evaluation gate that's queued for tomorrow. An Ollama-based host-side eval can swap **completion models** freely (they all expose the same chat API), but it can't swap the **embedder** without re-baking the seed corpus — every existing embedding was produced by `qwen3-0.6` and a different embedder's vectors won't align with them in cosine space. So U16's quirks-portable check has two halves:

1. **Completion winner ports cleanly on-device** — same prompt, same template, no per-model bespoke tuning. This is the easy axis.
2. **Embedder swap is harder than completion swap** — it requires re-baking all seed embeddings AND re-tuning `defaultMinScore` to match the new score distribution. Worth doing in the writeup ("the embedder slot is more locked-in than it looks") but probably not before the demo.

Cross-references:
- [`lib/services/cactus_service.dart`](../blob/main/lib/services/cactus_service.dart) — the two `CactusLM` instances + the slug pin
- [`_docs/notes/cactus-sdk-quirks.md`](../blob/main/_docs/notes/cactus-sdk-quirks.md) — the other Cactus SDK seam (`Supabase.getModel` firing regardless of telemetry flag)
- [`_docs/plans/001-feat-mesh-rag-demo.md`](../blob/main/_docs/plans/001-feat-mesh-rag-demo.md) §U16 — the alternate-model evaluation gate this constrains

---

## Update — filed upstream (2026-05-26)

Filed three issues against the cactus-flutter SDK that together cover the seams we hit here:

- **cactus-compute/cactus-flutter#33** — `isTelemetryEnabled` is ignored by `getModel` / `fetchModels` / `fetchVoiceModels`. The flag honors `sendLogRecord` and `registerDevice` but not the model-catalog functions, so opt-out apps still phone home on every boot. *Security-relevant; clean 3-line fix.*
- **cactus-compute/cactus-flutter#34** — chat-tuned slugs (`qwen3-1.7`, `qwen3-0.6`) accept `embed()` and fail at runtime with cryptic code `-2` instead of refusing at slug registration. *This is the bug that took us a half-day to diagnose on 2026-05-23.*
- **cactus-compute/cactus-flutter#35** — the purpose-built `qwen3-embedding-0.6` slug can't be resolved by the Flutter SDK 1.3.0 catalog despite being in the engine README. *This is why we ship a chat-tuned model in the embedder slot.*

#34 and #35 together explain the whole two-model story above: #34 is why we couldn't share one model, #35 is why the embedder is `qwen3-0.6` rather than the purpose-built `qwen3-embedding-0.6`. If either lands, the cosine-distribution side effects we tune around (`defaultMinScore=0.3`, the entity-overlap gate, the case-sensitive title-case normalization) might relax considerably.
